src/scales/scale.time.js

```diff
@@ -404,25 +402,27 @@ function ticksFromTimestamps(values, majorUnit) {
 	return ticks;
 }
 
-function determineLabelFormat(data, timeOpts) {
-	var i, momentDate, hasTime;
-	var ilen = data.length;
+/**
+ * Return the time format for the label with the most parts (milliseconds, second, etc.)
+ */
+function determineLabelFormat(timestamps) {
+	var presets = adapter.presets(true);
+	var ilen = timestamps.length;
+	var i, ts, hasTime;
 
-	// find the label with the most parts (milliseconds, minutes, etc.)
-	// format all labels with the same level of detail as the most specific label
 	for (i = 0; i < ilen; i++) {
-		momentDate = momentify(data[i], timeOpts);
-		if (momentDate.millisecond() !== 0) {
-			return 'MMM D, YYYY h:mm:ss.SSS a';
+		ts = timestamps[i];
+		if (ts % INTERVALS.second.size !== 0) {
+			return presets.millisecond;
 		}
-		if (momentDate.second() !== 0 || momentDate.minute() !== 0 || momentDate.hour() !== 0) {
+		if (ts % INTERVALS.day.size !== 0) {
```

Comment:
This will not work if in a timezone other than `UTC`.
Checking against `INTERVALS.hour.size` would work when tz offset is even hour, but will fail to detect time if at `hour` level (eg. '2018-01-08 04:00:00Z').
So in addition to `long presets`, there would needs to be a `hasTime` in adapter.

Reply:
Thanks @kurkle. Per my previous comment, I would instead get rid of this whole logic.

Reply:
`adapter.startOf(ts, 'day') !== ts`, to detect if ts has time information, works as expected 👍